test/commands/extended/get_inputs_test.py

```diff
@@ -818,3 +867,107 @@ def mock_address_generator(ag, start, step=1):
+
+    # To keep the unit test nice and speedy, we will mock the address
+    # generator.  We already have plenty of unit tests for that
+    # functionality, so we can get away with mocking it here.
+    # noinspection PyUnusedLocal
+
+    def mock_address_generator(ag, start, step=1):
+      # returning up to 3 addresses, depending on stop value
+      for addy in [self.addy0, self.addy1, self.addy2][start::step]:
+        yield addy
+
+    with mock.patch(
+        'iota.crypto.addresses.AddressGenerator.create_iterator',
+        mock_address_generator,
+    ):
+      with mock.patch(
+        'iota.commands.core.GetBalancesCommand._execute',
+        mock_get_balances_execute,
+      ):
+        response = self.command(
+          seed  = Seed.random(),
+          start = 1,
+          stop = 3,
+        )
+
+    self.assertEqual(len(response['inputs']), 2)  # 3 - 1 = 2 addresses expected
+    self.assertEqual(response['totalBalance'], 22)
+
+    input0 = response['inputs'][0]
+    input1 = response['inputs'][1]
+    self.assertIsInstance(input0, Address)
+    self.assertIsInstance(input1, Address)
+    self.assertEqual(input0, self.addy1)
+    self.assertEqual(input1, self.addy2)
+    self.assertEqual(input0.balance, 11)
+    self.assertEqual(input0.key_index, 1)
+    self.assertEqual(input1.balance, 11)
+    self.assertEqual(input1.key_index, 2)
+
+
+  def test_security_level(self):
+    """
+    Testing GetInputsCoommand  with selected security_levels
+    with and without `stop` parameter to cover both branches in the command
+    """
+    def invoke_cmd(seed, stopYN, securityLevel):
+      self.adapter.seed_response('getBalances', {
+        'balances': [86],
+      })
+      # ``getInputs`` uses ``findTransactions`` to identify unused
+      # addresses.
+      # noinspection SpellCheckingInspection
+      self.adapter.seed_response('findTransactions', {
+        'hashes': [
+          TransactionHash(
+            b'TESTVALUE9DONTUSEINPRODUCTION99999YFXGOD'
+            b'GISBJAX9PDJIRDMDV9DCRDCAEG9FN9KECCBDDFZ9H'
+          ),
+        ],
+      })
+      self.adapter.seed_response('findTransactions', {
+        'hashes': [],
+      })
+      if stopYN:
+        ret = GetInputsCommand(self.adapter)(
+              seed=seed,
+              stop=1,
+              securityLevel=securityLevel,
+            )
+      else:
+        ret = GetInputsCommand(self.adapter)(
+          seed=seed,
+          securityLevel=securityLevel,
+        )
+      return ret
+
+    # one address with index 0 for selected security levels for the random seed.
+    # to check with respective outputs from command
+    seed = Seed.random()
+    security_levels_to_test = [1, 2] # at least one is non-default
+    addrs = {l: AddressGenerator(seed, l).get_addresses(0)[0] for l in security_levels_to_test}
+
+    for securityLevel in security_levels_to_test:
+      for stop in [True, False]:
+        response = invoke_cmd(seed, stop, securityLevel)
```

Comment:
🤔 I like what you're going for here, but I am concerned that making the unit test more DRY incurs too high of a cost:

- If any assertion fails, the test stops, which limits the amount of information available for diagnosing test failures.

  For example, suppose the test fails when `stop = False` and `securityLevel = 1`.  As is, we can't be sure if _both_ conditions are required to cause the test failure, or if _either_ will.  If we split this into multiple tests, then we can see all of the different scenarios that cause pass/fail results, which will help us to look for patterns.
- It becomes more difficult to use the test as documentation; developers have to invest significant effort into understanding the unit test, before they can use it to understand the intended behaviour.
- Seeing conditionals and nested loops in a unit test makes me a bit nervous; this is where the complexity of the functionality reaches a point where I'd recommend we add some unit tests — except then we'd have to write tests for our unit tests!

Reply:
actually, I agree with you: is is unnecessary complication which makes me nervous too :) Replaced by two unit tests, both for non-default `security_level` value